mount/mount_id_mapped.go

```diff
@@ -0,0 +1,131 @@
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid container id for user namespace remapping")
+	}
+	hID, err := strconv.ParseUint(parts[1], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid host id for user namespace remapping")
+	}
+	size, err := strconv.ParseUint(parts[2], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, errors.Wrapf(err, "invalid size for user namespace remapping")
+	}
+
+	return []syscall.SysProcIDMap{
+		{
+			ContainerID: int(cID),
+			HostID:      int(hID),
+			Size:        int(size),
+		},
+	}, nil
+}
+
+func mountIDMapped(target string, pid int) (err error) {
+	var (
+		path       string
+		attr       Attr
+		userNsFile *os.File
+		targetDir  *os.File
+	)
+
+	path = fmt.Sprintf("/proc/%d/ns/user", pid)
+	if userNsFile, err = os.Open(path); err != nil {
+		return errors.Wrapf(err, "Unable to get user ns file descriptor for - %s", path)
+	}
+
+	attr.attrSet = MOUNT_ATTR_IDMAP
+	attr.attrClr = 0
+	attr.propagation = 0
+	attr.userNs = uint64(userNsFile.Fd())
+
+	defer userNsFile.Close()
+	if targetDir, err = os.Open(target); err != nil {
+		return errors.Wrapf(err, "Unable to get mount point target file descriptor - %s", target)
+	}
+
+	defer targetDir.Close()
+	return SetAttr(int(targetDir.Fd()), "", unix.AT_EMPTY_PATH|AT_RECURSIVE,
+		&attr, uint(unsafe.Sizeof(attr)))
+}
+
+// MapMount applies GID/UID shift according to gidmap/uidmap for target path
+func MapMount(uidmap string, gidmap string, target string) (err error) {
+	const (
+		userNsHelperBinary = "/bin/true"
+	)
+	// TODO: Avoid dependency on /bin/true or do in a completely different way
+	// Currently there is no way to pass idmapping directly to mount_setattr,
+	// this is not very convenient from the container runtime point of view.
+	// The id remapping procedure should be done in containerd, due to we have
+	// old approach that use recursive chown.
+	// Maybe it is necessary to think about moving of container rootfs ownership
+	// adjustment to runc due to runc has information about container user namespace.
+	// But personally I think that it would be better to add possibility to call
+	// mount_setattr with explicit id mappings and leave container runtime components
+	// responsibilities unchanged.
+	cmd := exec.Command(userNsHelperBinary)
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWUSER,
+	}
+
+	if cmd.SysProcAttr.UidMappings, err = parseIDMapping(uidmap); err != nil {
+		return err
+	}
+	if cmd.SysProcAttr.GidMappings, err = parseIDMapping(gidmap); err != nil {
+		return err
+	}
+
+	if err = cmd.Start(); err != nil {
+		return errors.Wrapf(err, "Failed to run the %s helper binary", userNsHelperBinary)
+	}
+
+	defer func() {
+		if err = cmd.Wait(); err != nil {
+			errors.Wrapf(err, "Failed to run the %s helper binary", userNsHelperBinary)
+		}
+	}()
+	if err = mountIDMapped(target, cmd.Process.Pid); err != nil {
```

Comment:
I'm wondering if we can do this before calling `task.Start` instead of trying to do it here with a helper binary.

Reply:
> It seems like the only reason this only works on overlay is because it is the only snapshotted that has been modified to add the fake contained-specific mount options (as in, these are not kernel mount opts).

Yes, this is true. But currently the only xfs, ext4 and fat filesystems support idmapped mount points. Btrfs is in progress. Even support of overlayfs is currently not upstream of Linux kernel. I think that Christian Brauner can provide more details here. 
I've performed testing using custom Linux kernel with integrated support of idmapped mounts for overlayfs.

> Meanwhile it seems like it should be possible to perform the mapping on (theoretically) any mount.
Yes, but not for all filesystems idmapped mounts are necessary. For example currently fuse-overlayfs has own solution to remap GID/UID for container and it is implemented in user space in scope of fuse-overlayfs. Maybe I am wrong but for me it looks reasonable to have GID/UID remapping functionality inside particular snapshotter implementation.

> It also seems like such mappings should be performed on the pid returned by task.Pid() instead of creating a separate helper to hold the namespace.
This would be good solution. But, please correct me if I am wrong here: we can call task.Pid() only after task.Start(), so we will have Pid after the process in container is in running state. But in this case it becomes possible for container to access rootfs that doesn't have idmapped mounts enabled yet. I think we shouldn't change GID/UID mapping in container that already have running process. Looks like this mapping for rootfs mount point should be done before task.Start().

Currently there is no way to provide idmappings directly to mount_setattr(), the only /proc/{pid}/gid_map and /proc/{pid}/uid_map can be used, and we have to create tmp user namespace to configure rootfs mount point.

For example early version of shiftfs have an ability to provide gid_map and uid_map directly to mount syscall. Fuse-overlayfs also use similar approach:
`$ fuse-overlayfs -o uidmapping=0:10:100:100:10000:2000,gidmapping=0:10:100:100:10000:2000,lowerdir=lowerdir/a:lowerdir/b,upperdir=up,workdir=workdir merged`

I think we should have a discussion regarding this, thank you very much for comments.

Reply:
side note: btrfs is supported in 5.15

Reply:
> I've performed testing using custom Linux kernel with integrated support of idmapped mounts for overlayfs.

What's the future of this PR if there's not overlayfs support for idmapped mounts in the upstream kernel? I suppose we should wait until that support is actually merged to continue with this PR (for the overlayfs part unless).

Reply:
Yes, overlayfs is unsupported by upstream kernel currently, and regarding overlayfs part you point reasonable. Let's wait. I'll try to ask Christian regarding the state to overlayfs. Thanks!

Reply:
Also another minor wrench here is we most likely want to remap the whole rootfs of the container, not just the CoW layer.

Reply:
> Also another minor wrench here is we most likely want to remap the whole rootfs of the container, not just the CoW layer.

Sorry, but why do you think that it is applied to CoW layer only. CoW layer, if we can call it so, for overlayfs is "upperdir". Merged folder itself is just a union mount of all the layers. The GID/UID shift is applied to "merged" folder itself, this is exactly rootfs. Overlayfs should transparently translate idmapped mounts GID/UIDs shift. As I understand it is not necessary to apply idmapped mount separately to every layer of overlayfs.

Or maybe I've misunderstood you, could you please clarify, how do you see it, and what issues do you expect here?

Thanks!

P.S.
Sorry guys, I was very busy with another tasks, so I plan to continue this work since next week more actively.

Reply:
I'm referring to other mounts, such as bind mounts from the host fs.

Reply:
Ok, I see. 
I think this is mostly related to the volume mounts, due to I don't see any reason to apply idmappings in other cases.

So, two cases for idmapped mounts:
1. mount of container rootfs
2. volume mounts in case of underlying filesystem support idmapped mounts

Do you expect to see something like automatic apply of idmapped mount in case of container with user namespace enabled has volume mounts?

Thanks!

Reply:
Yeah, will have to think about this.
If the snapshotted is handling the idmapping then no other mounts can be mapped (as I understand it).
Or we can just support idmapping at the snapshotted level or now and later when this is more ubiquitously supported move the option elsewhere for people who really want to map the entire resulting rootfs.

Reply:
Thanks! Yes, will also check for more cases. For now, maybe we need to do like this:

Reply:
Apparently iPad does not like the word "snapshotter"

Reply:
To follow up on this, I think idmapping volumes should be out of scope for containerd.

In these years, the mapping to be done in a volume was added to the mount section of the runtime-spec: https://github.com/opencontainers/runtime-spec/pull/1143. So, what we should do here is to pass the mapping to the OCI runtime. runc then will do the idmap mounts when this PR is merged: https://github.com/opencontainers/runc/pull/3717 (crun is doing the same too, it is already merged IIRC).

And with kubernetes 1.27, we modified the CRI so it includes the mappings on each mount message. Therefore, the kubelet will send the mappings needed for each mount, we should just receive them and send it to the OCI runtime.

The containerd part for handling volumes with idmap doing what I've just described is simple, I've opened a PR a few weeks ago with it: https://github.com/containerd/containerd/pull/8287 (and manually tested end to end, from k8s to runc and the kernel :)).